Initiate gossip_queries

bmancini55 opened the pull request:
Part 1 for initiating gossip_queries for #615.  This PR includes:

* Adds feature flag
* Adds msg types to wire enum
* Adds SendMessageEvents
* Adds message handlers to RoutingMessageHandler
* Implements querying logic in NetGraphMsgHandler

julianknutsen commented:
I would suggest basing your peer_handler changes on top of https://github.com/rust-bitcoin/rust-lightning/pull/714

Major restructure that will save you some merge headache down the road.

FYI: @ariard @arik-so

TheBlueMatt commented on lightning/src/ln/msgs.rs:
```diff
@@ -826,6 +826,33 @@ pub trait RoutingMessageHandler : Send + Sync {
 }
 
+
+/// A trait that describes an object that can receive gossip queries messages
```
I think we should just integrate this into RoutingMessageHandler. Indeed, it would make RoutingMessageHandler much more complex, but that's the way the protocol went, and the existing functions are not very useful without this, too. That would also create significantly less conflicts with other work, as Julian notes.

bmancini55 replied:
Sounds good, I've been thinking the same thing. Thanks!

TheBlueMatt reviewed (commented):
Structure looks good mod one note, would be nice to get some tests at least of the implementation of the message handler for NetGraphMsgHandler.

bmancini55 commented:
> I would suggest basing your peer_handler changes on top of #714

@julianknutsen, thanks I'll check it out.

ariard commented on lightning/src/util/events.rs:
```diff
@@ -345,6 +345,22 @@ pub enum MessageSendEvent {
+		/// The node_id of the node which should receive this message
+		node_id: PublicKey,
+		/// The message which should be sent.
+		msg: msgs::QueryShortChannelIds,
```
I think a worthy follow-up would be to coordinate better between query-gossips-at-connection and query-gossips-when-network-view-is-bad, AFAICT this patchset aims to query for each connection, which might be too much bandwidth hungry. I was thinking you could have a `timer_refresh_network_view_with_queries` in NetGraphMsgHandler on the model of PeerHandler::timer_tick_occured()

bmancini55 replied:
Ah yes, sorry about that. I was doing some dev testing and left code that performs a sync on each connection. I still need to clean up the PeerHandler code to respect `should_request_full_sync`.  

Thanks, I'll check out `PeerHandler::timer_tick_occured()`.  For periodic anti-entropy, it might be nice to connect to a random node from the network view and attempt a sync.

ariard reviewed (commented):
Overall looks good ! Mainly see my point on DisconnectPeer, I would be more liberal with gossips queries errors from remote peers, as severing connections have consequence for channel processing. Also, on those cases, BOLT 7 recommendation is a MAY, so we really have room here. I wouldn't be surprised this protocol to be a bit buggy in other implementations and we should avoid error contamination.

bmancini55 commented on lightning/src/routing/network_graph.rs:
```diff
@@ -239,6 +281,331 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+		if task.is_none() {
+			return Err(LightningError {
+				err: String::from("Received unknown reply_channel_range message"),
+				action: ErrorAction::DisconnectPeer { msg: None },
```
I've changed all validation failures to now return an IgnoreError instead of DisconnectPeer.

bmancini55 commented on lightning/src/routing/network_graph.rs:
```diff
@@ -239,6 +281,331 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+	pub number_of_blocks: u32,
+	/// The height of the first block received in a reply. This value
+	/// should be less than or equal to the first_blocknum request in
+	/// first_blocknum.
```
Fixed "in QueryChannelRange".

The responder is free to send a sequence of replies that cover a superset of the queried block range, so long as the first reply overlaps the start and the last reply overlaps the end.

bmancini55 commented:
Updates: 
* Changed the handler methods to use IgnoreError instead of DisconnectPeer
* Added tests for gossip related methods in `NetGraphMsgHandler`.  Some are broader scoped than unit tests since I wanted to ensure the task setup wasn't faked.  Some of the test code is also a bit repetitive, I like making test code as straight forward as possible. Happy to try to reduce some of the repetitiveness if that fits the project style better.

More to follow!

ariard commented:
@bmancini55 Is this ready for another round of review :) ?

bmancini55 commented:
> @bmancini55 Is this ready for another round of review :) ?
> 
> EDIT: I'll have a look on the new tests anyway

@ariard I'll be adding more commits this week, so not quite there yet for a complete look. The tests should be good to review though, thanks!

ariard commented on lightning/src/routing/network_graph.rs:
```diff
@@ -1765,4 +2154,647 @@ mod tests {
+		let first_blocknum = 0;
+		let number_of_blocks = 0xffff_ffff;
+
+		// When no active query exists for the node, it should send a query message and generate a task
```
I wonder if the active query be time-out at some point in case of irresponsive peer ?

bmancini55 replied:
Definitely agree on this point.  Also need to clean up tasks on peer disconnection.

ariard commented on lightning/src/routing/network_graph.rs:
```diff
@@ -237,6 +282,349 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+		let mut query_range_tasks_lock = self.query_channel_range_tasks.lock().unwrap();
+		let query_range_tasks = &mut *query_range_tasks_lock;
+		if query_range_tasks.contains_key(their_node_id) {
+			return Err(LightningError {
```
If it's a caller requirement to track channel_range query I think this might be a hint of a higher logic bug. Though I don't have a clear model of this method usage.

bmancini55 replied:
My thinking is that this will be called from peer_handler after init negotiation and `should_request_full_sync` returns true. This guard prevents multiple in-flight `query_channel_range` messages which is spec non-compliant. This could be changed to queue channel range queries but I thought outright rejection was a simpler for now.  

Relatedly, my plan is to make `full_syncs_requested` decrement on any failure in `handle_reply_channel_range` so that a full sync can be requested with the next peer.

ariard replied:
Right, we can switch it later to an assert when higher sync logic is implemented.

ariard commented on lightning/src/routing/network_graph.rs:
```diff
@@ -98,6 +121,28 @@ impl<C: Deref, L: Deref> NetGraphMsgHandler<C, L> where C::Target: ChainWatchInt
+
+	/// Enqeues a message send event for a batch of short_channel_ids
+	/// in a task.
+	fn send_channels_query(&self, task: &mut GossipQueryShortChannelIdsTask) {
```
nit: Maybe a better name would be `finalize_query_short_ids`, as otherwise you may think it can used for `SendChannelRangeQuery`

ariard commented on lightning/src/routing/network_graph.rs:
```diff
@@ -237,6 +282,349 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+		// we received information as well.
+		if !msg.full_information && msg.short_channel_ids.len() == 0 {
+			query_range_tasks.remove(their_node_id);
+			return Err(LightningError {
```
I don't have test wreckage on this mutation, shouldn't this be covered by "Test receipt of a reply that indicates the remote node does not maintain up-to-date information for the chain hash" ?

```
diff --git a/lightning/src/routing/network_graph.rs b/lightning/src/routing/network_graph.rs
index 8c98e0e8..269df229 100644
--- a/lightning/src/routing/network_graph.rs
+++ b/lightning/src/routing/network_graph.rs
@@ -380,10 +380,11 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
                // we received information as well.
                if !msg.full_information && msg.short_channel_ids.len() == 0 {
                        query_range_tasks.remove(their_node_id);
-                       return Err(LightningError {
-                               err: String::from("Received reply_channel_range with no information available"),                                                                                                  
-                               action: ErrorAction::IgnoreError,
-                       });
+                       return Ok(())
+                       //return Err(LightningError {
+                       //      err: String::from("Received reply_channel_range with no information available"),                                                                                                  
+                       //      action: ErrorAction::IgnoreError,
+                       //});
                }

                // On the first message...
```

ariard commented on lightning/src/routing/network_graph.rs:
```diff
@@ -237,6 +282,349 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+			// not, the message is malformed and we return an error.
+			if msg.first_blocknum > task.first_blocknum {
+				query_range_tasks.remove(their_node_id);
+				return Err(LightningError {
```
Same, not covered by "Test receipt of a reply that has a first_blocknum that is above the first_blocknum requested in our query" ?

```
diff --git a/lightning/src/routing/network_graph.rs b/lightning/src/routing/network_graph.rs
index 8c98e0e8..f708a3e2 100644
--- a/lightning/src/routing/network_graph.rs
+++ b/lightning/src/routing/network_graph.rs
@@ -394,10 +394,11 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
                        // not, the message is malformed and we return an error.
                        if msg.first_blocknum > task.first_blocknum {
                                query_range_tasks.remove(their_node_id);
-                               return Err(LightningError {
-                                       err: String::from("Failing reply_channel_range with invalid first_blocknum"),
-                                       action: ErrorAction::IgnoreError,
-                               });
+                               return Ok(())
+                               //return Err(LightningError {
+                               //      err: String::from("Failing reply_channel_range with invalid first_blocknum"),
+                               //      action: ErrorAction::IgnoreError,
+                               //});
                        }
 
                        // Validate that the receive
```

ariard commented on lightning/src/routing/network_graph.rs:
```diff
@@ -1765,4 +2154,647 @@ mod tests {
+		let node_privkey_1 = &SecretKey::from_slice(&[42; 32]).unwrap();
+		let node_id_1 = PublicKey::from_secret_key(&secp_ctx, node_privkey_1);
+
+		let chain_hash = genesis_block(Network::Testnet).header.block_hash();
```
You initiate your task chain hash with `Network::Testnet` but further down you sound to test your error handling with `Network::Bitcoin`, thus likely not hitting the right error as expected.

Maybe you should assert that error string is matching the expected one, even if it's prone to break that's a worthy hardening.

ariard reviewed (commented):
So far the code is pretty well-documented and coverage sounds to be exhaustive (missing `handle_query_channel_range`/`handle_query_short_channels_ids` IIUC)

Still have a look on the mutation I provided and the point on the task chain hash, I'm not sure if it's correct.

If you have a follow-up in mind, you're welcome to track them in an appropriate issue, let avoids this PR to become obese.

bmancini55 commented:
@ariard :facepalm: major apologies for that sloppiness!! I have fixed the issues with `chain_hash` and reply fields.  Also took your advice to harden by asserting the message. I added tests for `handle_query_channel_range` and `handle_query_short_channels_ids` and renamed `send_channels_query` as requested.

> If you have a follow-up in mind, you're welcome to track them in an appropriate issue, let avoids this PR to become obese. 

If you're ok with it, this PR can be limited to:
* add gossip_queries feature (will change so it is not a default)
* wire enum changes for gossip queries messages
* add gossip_queries methods to `RoutingMessageHandler`
* implement handler methods in `NetGraphMsgHandler` + tests

Subsequent PRs can:
* initiate gossip_query sync in peer_handler on peer connection 
* respond to peer's gossip queries
* clean up orphaned/timed out tasks
* initiate periodic anti-entropy synchronization

If that sounds good, I will remove temporary code from peer_handler, squash commits appropriately, and flip this to ready for review.

ariard commented:
> If that sounds good, I will remove temporary code from peer_handler, squash commits appropriately, and flip this to ready for review.

SGTM with proposed scope!

ariard commented on lightning/src/routing/network_graph.rs:
```diff
@@ -243,6 +288,380 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+
+		// If the remote node does not have up-to-date information for the
+		// chain_hash they will set full_information=false. We can fail
+		// the result and try again with a different peer.
```
AFAIU this protocol, lack of information fullness doesn't imply you've not received any `channel_announcement`/`channel_update`. Function comment should warn initial caller to update it set of scids before to proceed with another peer.

ariard reviewed (commented):
Sounds almost good to me, only call for better naming/comments. Thanks for the great code comments, it make the understand of this fuzzy query protocol easier.

Note to myself : review again test coverage at next round.

julianknutsen commented:
@ariard What was the final conclusion on the dependency between this and #691? These patches look to be based on master and didn't see a resolution in any of the comments about the ordering.

bmancini55 commented:
@julianknutsen @ariard I'm happy to rebase this PR if desired.  The changes in peer_manager are fairly minimal for this PR. 

@ariard, I'm looking over your comments and will make those changes shortly. Thanks!

ariard commented:
@bmancini55 It can go first, it's more near-merge mature than #691 :) 

@julianknutsen What's the real issue with the ordering ? AFAICT, there is only one _potential_ conflicting file `lightning-net-tokio/src/lib.rs`and in anycase it would be a minor one ? I might be wrong, in that case may you point the conflict and how much complexity it would be to fix it if #691 go second ?

bmancini55 commented on lightning/src/routing/network_graph.rs:
```diff
@@ -243,6 +288,380 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+		Ok(())
+	}
+
+	fn handle_query_channel_range(&self, _their_node_id: &PublicKey, _msg: &QueryChannelRange) -> Result<(), LightningError> {
```
Added a comment to both methods.

bmancini55 commented on lightning/src/routing/network_graph.rs:
```diff
@@ -243,6 +288,380 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+/// added to it while we are waiting for a reply_short_channel_ids_end.
+/// The pending short_channel_ids will be chunked and sent to the remote
+/// peer.
+pub struct GossipQueryShortChannelIdsTask {
```
Comment added, hopefully that clarifies the state.

bmancini55 commented on lightning/src/routing/network_graph.rs:
```diff
@@ -243,6 +288,380 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+		if task_received_first_block.is_none() {
+			// The replies must include our range but can be a superset of the
+			// queried block range. We start by validating the reply's
+			// first_blocknum is less than or equal to the start of our query.
```
Clarified the comment here

bmancini55 commented on lightning/src/routing/network_graph.rs:
```diff
@@ -18,24 +18,38 @@ use bitcoin::hashes::Hash;
+
+/// Maximum number of reply_channel_range messages we will allow in
+/// in reply to a query_channel_range.
+const MAX_REPLY_CHANNEL_RANGE_PER_QUERY: usize = 250;
```
They are not defined in BOLT 7. I updated the comments with more information. 8000 is used in other implementations as the raw batch size to fit inside the 65535-byte message limit. I can fill you in on max replies over email if you want more details.

ariard replied:
Good, the 65535 as an upper bound makes sense to me.

TheBlueMatt commented on lightning/src/ln/msgs.rs:
```diff
@@ -824,6 +824,27 @@ pub trait RoutingMessageHandler : Send + Sync {
 	fn should_request_full_sync(&self, node_id: &PublicKey) -> bool;
+	/// Queries a peer for a list of channels with a funding UTXO in the requested
+	/// chain and range of blocks.
+	fn query_channel_range(&self, their_node_id: &PublicKey, chain_hash: BlockHash, first_blocknum: u32, number_of_blocks: u32) -> Result<(), LightningError>;
```
I think these fn interfaces can be cut back a bit - at least chain_hash can be dropped because `NetworkGraph` objects only actually support one chain (though maybe we should track it in the `NetworkGraph`!) and I'm not sure exactly how this would be used so not sure why we need `number_of_blocks` - can we just always set it to MAX_INT? It may help to document exactly how these intend to be used, though I've left some further motivating comments below.

bmancini55 replied:
There are three use cases we may be interested in:
1. A full sync 
2. A partial sync
3. Perform an SCID query if we receive an unknown channel_update

Focusing on the full sync, we can drop the `first_blocknum` and `number_of_blocks` arguments and simplify the interface to a single method `perform_routing_sync(&self, their_node_id: &PublicKey)` or similarly named. This could initiate the message flow that covers both channel discovery and request of info. It will depend on how we want to handle state (more on that below).

ariard replied:
I think you may want to keep `number_of_blocks`, you might have clients with limited storage and thus building their network map iteratively (request channel announcement/updates for a range, discard not-interesting channels, move forward). I agree on the NetworkGraph+chain_hash, though it can be done as a follow-up.

TheBlueMatt replied:
Right, I guess my thinking here is in part because we haven't implemented the peer_handler part here, and I don't know how interesting it would be to have a generic API which allows all kinds of requests if we only ever make one or two request types in our own code - my gut says probably not so interesting.

TheBlueMatt commented on lightning/src/routing/network_graph.rs:
```diff
@@ -243,6 +290,414 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+
+		// First we obtain a lock on the task hashmap. In order to avoid borrowing issues
+		// we will access the task as needed.
+		let mut query_range_tasks = self.chan_range_query_tasks.lock().unwrap();
```
I'm not convinced we need the state tracking/validation at all - we query a peer for all the channel IDs in a given range, and it feeds us back channel IDs, we then do a ton of validation to make sure they line up with exactly the query we sent before asking for the channel announcements. In the case the peer is malicious and trying to make us waste upload bandwidth by sending more queries, they could tell us about fake channels in the right range (or just send us pings), and in the case they're buggy we should find that out when they send us the channel announcements in question and they fail to validate.

ariard replied:
Yeah I did have the same wonder about the state tracking purpose : https://github.com/rust-bitcoin/rust-lightning/pull/736#discussion_r514370640, my conclusion was that it's required by the way the spec is structured. Now does the overall gossip spec is DoS safe ? I don't think so if you assume disk-IO one on `channel_announcement` un-bounded reception.

I don't think there is a necessity that out-of-range channel announcements fail to validate if we have the blockstate yet available ?

TheBlueMatt commented on lightning/src/routing/network_graph.rs:
```diff
@@ -243,6 +290,414 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+				let task = query_range_tasks.get_mut(their_node_id).unwrap();
+				let mut short_channel_ids = Vec::new();
+				std::mem::swap(&mut short_channel_ids, &mut task.short_channel_ids);
+				self.query_short_channel_ids(their_node_id, task.chain_hash, short_channel_ids)?;
```
Why not send the queries as we go? With that I think we can drop all of the state tracking here, which would clean up the code a ton and make reasoning about state and DoS issues much simpler.

bmancini55 replied:
I don't think we can make the SCID queries entirely stateless. The spec only allows a single query of each type to be in-flight at a time. A large channel range query will stream multiple batches of SCIDs. Because we can only have a single SCID query in-flight, we need to store the subsequent batches and wait until we get an `_end` message before we can initiate another SCID query. For example:

```
|------ query_channel_range               ----->|
|<----- reply_channel_range (batch 1)     ------|
|<----- reply_channel_range (batch 2)     ------|
|<----- reply_channel_range (batch n)     ------|
|                                               |
|------ query_short_channel_ids (batch 1) ----->|
|<----- many chan_ann, etc...             ------|
|<----- reply_short_channel_ids_end       ------|
|                                               |
|------ query_short_channel_ids (batch 2) ----->|
|<----- many chan_ann, etc...             ------|
|<----- reply_short_channel_ids_end       ------|
|                                               |
|------ query_short_channel_ids (batch n) ----->|
|<----- many chan_ann, etc...             ------|
|<----- reply_short_channel_ids_end       ------|
```
So at a bare minimum we'll need to serialize and defer SCID query events some how, even if we initiate the first SCID query as soon as receive the first batch.

The path I went down allows us to handle the three use cases described above, but is overkill if we only plan to support full syncs. I'm happy to come up with a few options to simplify things and report back!

TheBlueMatt replied:
Ugh. I wonder if we should just get the spec to relax the query limit for short_channel_ids data so that we can make all our responses completely stateless (modulo network buffers). The alternative would be to tightly limit the response set size and maybe even do requests one block at a time to make sure we can set a tight limit and not have to worry about it. I do think we can still cut back a bunch on the amount of state we track to enforce checks that aren't really DoS risks.

TheBlueMatt reviewed (commented):
Sorry about the delay, was off for the last week. I think we can cut back a lot of code here by being a bit more willing to respond to bogus-but-not-harmful data from peers.

ariard commented:
Thanks @julianknutsen for the information, added a reference to your comment in #707 backlog.

ariard commented:
Thanks for the great new comments at 6a59483.

@TheBlueMatt We might bound the size of the `scid` query after `reply_channel_range` batch completion to avoid upload bandwidth/disk-IO DoS ? Assuming we pick up a per-peer bound, we may use the historical average number of channel opening per-block on the last year as an heuristic, plus some safety margin. The hard limit is max_number_of_p2wsh_per_block. 

Also you assume a DoS adversary is able to gain our gossips-sync peer selection. We may recommend to pick them randomly to mitigate as a first step.

That said, I think we should defer this DoS discussion to future PRs, cutting in the implementation as required by the spec isn't going to solve the issue, as ultimately it's a question of receiving non-sanitized data from untrusted peers. IMO, we should first a) know the default bandwidth resources we assume on a RL client (but maybe configurable?) b) define how do we select gossip-sync peers c) allocate a gossip budget to each peer

TheBlueMatt commented:
> We might bound the size of the scid query after reply_channel_range batch completion to avoid upload bandwidth/disk-IO DoS ? 

I can't say I'm super concerned about upload bandwidth - assuming we use a similar approach to the existing map sync stuff where we only send such messages with low priority and let channel-related messages go first, its not any worse than what a peer can do with ping/pongs anyway. As for disk, for now our map is all in memory, though it would be something worth considering if we ever move to a disk-backed version.

> Also you assume a DoS adversary is able to gain our gossips-sync peer selection. We may recommend to pick them randomly to mitigate as a first step.

Maybe? I assume we'll want to sync with multiple peers at least, eventually doing some reconciliation. I think we can pretty tightly limit things and then we don't have to worry about it, though.

> That said, I think we should defer this DoS discussion to future PRs

I don't see why? We have to deal with it sooner or later, might as well get it right in the first go? I think it should be pretty doable to make things pretty cheap here.

bmancini55 commented:
> I wonder if we should just get the spec to relax the query limit for short_channel_ids data

It would simplify things on the query end. IIRC, it's a simple DoS prevention mechanism so the query recipient can throttle by taking time responding to a query.

> The alternative would be to tightly limit the response set size and maybe even do requests one block at a time to make sure we can set a tight limit and not have to worry about it.

One-by-one is a possibility, though our peers might not appreciate that haha. Alternatively, we could process just the first reply in a sequence. When the `scid` query completes for the first reply, we issue a new range query starting next in the sequence. The only state we would need in both of these options is the block height of where we are in the sync process.

To simplify as-is, we could prune the interface to only a sync method and use a single `SyncTask` struct to track completion of the range query and hold `ScidQueryEvent`s that get flushed in sequence. It's not substantially different than the current code, but simplifies some of the moving pieces.

Generally I wasn't overly concerned with state because we're only going to initiate a sync with a small number of peers and we have an upper-bound on allocation of 16MB (250 messages limit * ~8000 `scid` * 8-bytes per `scid`) before we abort and clean up. But I certainly understand a desire to go stateless from an allocation and complexity standpoint!

Regardless of state, I also agree that we could simplify `reply_channel_range` validation logic without loss of functionality. If a node is going to send bogus data, the sequence validation isn't going to help us. We'll still need to validate whether the responder failed the query and if the query is complete. If we would rather leave the validation in, I'm ok with that too haha

TheBlueMatt commented:
>  IIRC, it's a simple DoS prevention mechanism so the query recipient can throttle by taking time responding to a query.

Right, I see the motivation there, but it doesn't really work that way - in practice you're never going to implement it by not processing a new inbound message until you've finished uploading the responses from the previous message (and checking with the kernel that its already been sent?), so a peer can almost always send several in a row.

I think maybe the thinking might have been more along the lines of making sure other messages can be sent in between by encoding into the spec the idea of not having too much in-flight at once, though that *also* doesn't solve the problem (there's plenty of other issues around fresh announcements flooding message queues) and you really need a solution like ours where you have (effectively, in our case) a separate outbound message queue for "background" messages and one for channel messages.

I don't think its unreasonably to change the spec here, though other implementations may disagree. I'd be curious to know if any implementations actually have the ability to enforce the "one-in-flight" rule, anyway.

bmancini55 commented:
> in practice you're never going to implement it by not processing a new inbound message until you've finished uploading the responses from the previous message 

Sure, a recipient doesn't need to _fully_ process a message though. They can fail the query or disconnect if a query buffer is full (size of 1 according to the spec haha).

> I'd be curious to know if any implementations actually have the ability to enforce the "one-in-flight" rule, anyway.

I ran some tests on concurrent `query_short_channel_ids` messages sending

* Eclair disconnects you after 520 concurrent queries
* LND will time-delay handling/responding after the 10th concurrent query, I didn't encounter a disconnect limit :eyes:
* c-lightning will disconnect you on the second concurrent query

> I don't think its unreasonably to change the spec here, though other implementations may disagree. 

Agreed, it would simplify initiating both sync and anti-entropy queries for unknown `channel_update`. Query recipients will already need to consider DoS vectors. The spec is putting some of that burden on the query initiator as well.

Do you want me to flip this back to draft and modify down that path? If so, I'm thinking:
* simplify `RoutingMessageHandler` as discussed above
* modify `NetGraphMsgHandler` to remove state, queue `SendScidQueryEvent`s immediately from a `reply_channel_range`, and remove sequence validations for `handle_reply_channel_range`

ariard commented:
Just to be sure we're all on the same framework.

The DoS scenario we're concerned of is a memory exhaustion due to struct `ChanRangeQueryTask`:
* Our node is sending a `query_channel_range` of size X blocks
* The malicious query recipient send 250 (`MAX_REPLY_CHANNEL_RANGE_PER_QUERY`) messages * 8000 * 8-bytes, i.e 16MB and hold on sending a final `reply_channel_range` completing our queried range
* After some timer expiration or in concurrency, our node is starting the querying process with another peer, also malicious and repeating the step just above
* After connecting to Y query-withholding peers, our node crash OOM, Y * 16MB >= process memory hard limit

The proposed solution is to go stateless and remove `ChanRangeQueryTask`, thus issuing `query_short_channel_ids`. This implies a spec change to allow all implementations authorizing concurrent scid queries.

That said, we may have alternative DoS mitigation :
* add a timer `reply_time_bound` to `ChanRangeQueryTask`, store time when sending the query
* add a `timer_watch_range_queries()` to `NetGraphMsgHandler`, if a query is still pending after X seconds, discard the querying state, you might conserve scids received so far
* if your peer send `reply_channel_range` after the state cleanup, discard them
* ensure you don't open more than Y connections, where Y * 16MB == your_querying_state_memory_limit. When the first peer of Y expires its `reply_time_bound` replace it with a new one

I do agree that this mitigation is harder to implement correctly but AFAICT it comes without spec changes.

> Do you want me to flip this back to draft and modify down that path? If so, I'm thinking:

Can we go forward with proposed changes, even if spec doesn't change, it sounds we have concurreny margin with LND/Eclair ?

bmancini55 commented:
Thanks for the excellent overview @ariard!

> Can we go forward with proposed changes, even if spec doesn't change, it sounds we have concurreny margin with LND/Eclair ?

I think so, we have some maneuverability with that approach.

In the event we can't get the spec changed to multiple inflight `query_short_channel_ids`, we could lobby for a similar strategy with `query_channel_range`/`reply_channel_range` as `getblocks`. A broad query returns a single reply, and we must issue a subsequent query with the next range when we are ready.

And if that isn't acceptable, we can mimic `getblocks` flow by ignoring sequenced `reply_channel_range` by only processing the first `reply_channel_range` message we receive in response to a query.

ariard commented:
> And if that isn't acceptable, we can mimic getblocks flow by ignoring sequenced reply_channel_range by only processing the first reply_channel_range message we receive in response to a query.

Yes so we couple `query_channel_range`/`query_short_channel_ids` as a sequence and just keep a block hash pointer we increase after each sequence completion to sweep the full-block range we're interested with per peer. Sounds to me an acceptable worst-case if we don't get the spec update we want and have to hack back our stuff.

Feel free on the stateless approach, it has my Concept ACK :)

bmancini55 commented:
The stateless version has been pushed. Couple notes:

* The changes are as an additive commit. I figured it would be good to retain the history of a fully spec compliant implementation in the event we need to reintroduce anything. Also happy to rebase it if desired.
* The `sync_channel_range` method still retains the `chain_hash` argument. From what I could tell, `NetGraphMsgHandler` and `NetworkGraph` don't have a reference to a current chain. The only reference I could find in network_graph is `handle_channel_announcement` which uses the `chain_hash` from the received message. I'm open to suggestions on how you guys want me to handle this.
* I've removed the validations in `handle_reply_channel_range` against the range query. The handler now immediately echos the  SCIDS to the SCID query message

TheBlueMatt commented on lightning/src/routing/network_graph.rs:
```diff
@@ -243,6 +250,122 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+	/// will enqueue a query_scid message to request gossip messages for each
+	/// channel. The sync is considered complete when the final reply_scids_end
+	/// message is received, though we are not tracking this directly.
+	fn sync_channel_range(&self, their_node_id: &PublicKey, chain_hash: BlockHash, first_blocknum: u32, number_of_blocks: u32) {
```
Right, so this is still a bit verbose.

As for `chain_hash`, I think just make the `NetworkGraph` aware of the current chain hash and take it in the constructor or so. We probably want to do this anyway because we should check the chain_hash in `handle_channel_announcement` even if we don't have a `chain_access` filled in.

In general, though, I'm just not quite sure what the right answer is here - it seems wrong for the peer handler (something which is really meant to just be a dumb multiplexer of events and messages across handlers and peers) to know anything about the current chain height, either. Instead, a simple "peer has connected, their_node_id is X" API seems like what `PeerHandler` wants, though what we do from there is murky - making NetGraphMsgHandler aware of the current chain also sucks.

Do you think it makes sense to use that information to do a sync from all peers that starts from block `95th percentile of the channels we have`? It would be vulnerable to fake channel announcement attacks, but maybe that's OK? Obviously long-term we should be thinking about using set-reconciliation to fix this, but that's probably a ways off.

bmancini55 commented:
> a simple "peer has connected, their_node_id is X" API seems like what PeerHandler wants, though what we do from there is murky - making NetGraphMsgHandler aware of the current chain also sucks.

Thinking more on this, I agree that PeerHandler should invoke sync with only `their_node_id`. How NetGraphMsgHandler performs the sync is internal to NetGraphMsgHandler. This gives us flexibility if we want to change the implementation or make it an extensible strategy via an injected trait object. The latter could be used to create a chain-aware strategy without explicitly adding chain knowledge to NetGraphMsgHandler.

For a first pass though, I think a simple full sync for X number of peers is the best way to go.

TheBlueMatt commented:
> For a first pass though, I think a simple full sync for X number of peers is the best way to go.

Agreed. Presumably that is with the API changes I suggested at https://github.com/rust-bitcoin/rust-lightning/pull/736#discussion_r528216872, though? Or, at least, your comment seems somewhat orthogonal to those suggested changes - the current changes here require `PeerManager` to get smarter about the chain/graph sync state though in the long-term we want to get even dumber.

bmancini55 commented:
Sorry, should have been more clear about intent, but yes, I intend to modify to your suggested [comments](https://github.com/rust-bitcoin/rust-lightning/pull/736#discussion_r528216872). In summary:

* `sync_channel_range` signature becomes `fn sync_channel_range(&self, their_node_id: &PublicKey)` 
* Add `chain_hash` to `NetworkGraph` instead of supplying via `sync_channel_range` argument
* It might make sense to rename to `sync_routing_table` or `sync_routing_topology` as well

TheBlueMatt commented:
Cool! Sounds great.

bmancini55 commented:
Sorry for the delay, took some time off with the holiday and just got a chance to circle back to this. Latest update pushed as described above. 

A few notes about adding the `BlockHash` to `NetworkGraph` in a9c88c60a1d2f23fb5f74c192082329850eabd22.  

* This required added a `BlockHash` argument to the `new` method of `NetGraphMsgHandler` to support construction of the `NetworkGraph` (https://github.com/rust-bitcoin/rust-lightning/commit/a9c88c60a1d2f23fb5f74c192082329850eabd22#diff-e68ca216ace695426d8cc8a641448f25fb88c18f8d3149a879465ef8498b077cR91)
* For `NetworkGraph` serialization, I added `BlockHash` as the first piece of data. This seemed most logical to me since the other data is variable in length.  (https://github.com/rust-bitcoin/rust-lightning/commit/a9c88c60a1d2f23fb5f74c192082329850eabd22#diff-e68ca216ace695426d8cc8a641448f25fb88c18f8d3149a879465ef8498b077cR935)

Happy to modify this if needed.

TheBlueMatt commented on lightning/src/ln/msgs.rs:
```diff
@@ -827,6 +827,25 @@ pub trait RoutingMessageHandler : Send + Sync {
+	/// Initiates routing gossip sync by querying a peer to discover channels
+	/// and their associated routing gossip messages. This method will use a
+	/// sync strategy defined by the implementor.
+	fn sync_routing_table(&self, their_node_id: &PublicKey);
```
It may be useful to pass in the `Init` message here so that the NetGraphMessageHandler can analyze the InitFeatures. Same goes for `should_request_full_sync` - that way we can only return true in the second when the peer doesn't support gossip queries and send a query_channel_range when they do.

TheBlueMatt commented on lightning/src/ln/peer_handler.rs:
```diff
@@ -694,10 +694,11 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref> PeerManager<D
 					if msg.features.supports_data_loss_protect() { "supported" } else { "not supported"},
 					if msg.features.initial_routing_sync() { "requested" } else { "not requested" },
 					if msg.features.supports_upfront_shutdown_script() { "supported" } else { "not supported"},
+					if msg.features.supports_gossip_queries() { "supported" } else { "not supported" },
```
I presume we should call `sync_routing_table` at the bottom of this block where we call `peer_connected`?

TheBlueMatt reviewed (commented):
This is looking good. Mostly trivial nitpicky stuff here. I didn't test it against any live nodes, have you gotten a chance to?

bmancini55 commented on lightning/src/routing/network_graph.rs:
```diff
@@ -212,6 +217,126 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+			});
+		}
+
+		// Copy the SCIDs into a new vector to be sent in the SCID query
```
I don't see why we can't use move for the query messages, I used by-ref for consistency.

bmancini55 commented on lightning/src/routing/network_graph.rs:
```diff
@@ -212,6 +217,126 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+	/// Handling requests with first_blocknum very far away may trigger repeated
+	/// disk I/O if the NetworkGraph is not fully in-memory.
+	fn handle_query_channel_range(&self, _their_node_id: &PublicKey, _msg: &QueryChannelRange) -> Result<(), LightningError> {
+		// TODO
```
Indeed, want to outbound queries buttoned up before diving into responding to them.

bmancini55 commented:
> This is looking good. Mostly trivial nitpicky stuff here. I didn't test it against any live nodes, have you gotten a chance to?

Yep tested against LND with a minor modifications to `PeerManager` to call `sync_routing_table` when the peer connects, as you noted above.

On that, I was trying to keep the touches on `PeerManager` to a minimum and was planning to submit another PR to conditionally send the query or set the sync feature bit.  Should I just make those changes now so this PR can button up initiating queries?

bmancini55 commented:
New updates:
* changed handler methods to pass ownership of gossip messages
* passing `Init` message to `sync_routing_table`
* call `sync_routing_table` after receipt of `Init` from peer

ariard commented on lightning/src/routing/network_graph.rs:
```diff
@@ -1813,4 +1932,165 @@ mod tests {
+					0x0006e0_000000_0000, // 1760x0x0
+				],
+			});
+			assert!(result.is_ok());
```
Given `handle_reply_channel_range` doesn't verify anymore that scid replies are bounded in queried channel range is this test check still useful ?

bmancini55 replied:
Yep, the test is still valid, but the comment on line 1983 is outdated, will clean that up.

ariard commented on lightning/src/routing/network_graph.rs:
```diff
@@ -1813,4 +1932,165 @@ mod tests {
+	}
+
+	#[test]
+	fn handling_query_channel_range() {
```
What does this test and the following are checking ? Are they still accurate given last modifications ?

bmancini55 replied:
These two tests are for the stubbed method for handling inbound queries.  Since they're stubbed they just test that it results in an Err.

ariard replied:
Ah yes true.

ariard commented on lightning/src/ln/msgs.rs:
```diff
@@ -825,8 +825,33 @@ pub trait RoutingMessageHandler : Send + Sync {
+	/// Handles when a peer asks us to send routing gossip messages for a
+	/// list of short_channel_ids. There are potential DoS vectors when handling
+	/// inbound queries. Handling requests with first_blocknum very far away may
+	/// trigger repeated disk I/O if the NetworkGraph is not fully in-memory.
```
Alternatively, you can gather potential DoS vectors in "Implementor DoS Warnings" section around `RoutingMessageHandler` trait doc, with trade-offs if graph isn't in memory. I'm fine being in trait function docs, your call if you want to remove reference to `NetworkGraph`.

ariard reviewed (commented):
Sorry for delay, the stateless approach is so much simpler !

> What I'm doing is using should_request_full_sync to set the initialize_routing_sync feature bit and flag the peer as requesting a sync. Then once we receive the peer's Init message, we use that flag to control if sync_routing_table should be called.

IIUC, setting `initialize_routing_sync` on our `Init` message should trigger the receiving peer to send us a complete routing table dump, at least that the behavior recommended by the spec ? But how many implementations are still supporting this behavior, it sounds a huge upstream bandwidth leak ? I'm of the opinion to remove it completely and only rely on gossip queries.

Also have a log on lightning-dev meeting of yesterday : http://gnusha.org/lightning-dev/2020-12-07.log, we discussed both #804/#811
* it has been proposed to repurpose `full_information` of channel range query as a clear EOF, thus standardizing behavior among implementation
* we discussed introducing a new feature `no-zlib-support` thus easing discovery of raw encoding supporting peers for gossip queries

And we still have to make the stateless approach lawful by allowing concurrent gossip queries. 

What I'm thinking of is landing this PR once minor points are solved and then write down the spec updates to propose changes matching our requirement ?

bmancini55 commented:
> IIUC, setting initialize_routing_sync on our Init message should trigger the receiving peer to send us a complete routing table dump, at least that the behavior recommended by the spec ? But how many implementations are still supporting this behavior, it sounds a huge upstream bandwidth leak ? I'm of the opinion to remove it completely and only rely on gossip queries.

When we enable `gossip_queries` feature in our `init` message the remote peer will ignore the `initial_routing_sync` flag. I know LND dropped support for `initial_routing_sync` entirely regardless of whether `gossip_queries` is enabled. I haven't recently tested Eclair or CL to see if they still support `initial_routing_sync` when `gossip_queries` isn't negotiated.

For outbound sync, I'm in agreement that we should not set `inititial_routing_sync` and only rely on `gossip_queries`. This has the benefit of removing that conditional logic from `PeerManager` and removing `should_request_full_sync` from `RoutingMessageHandler`.  Sync can then be fully encapsulated inside the implementation of `RoutingMessageHandler`.

> Also have a look on lightning-dev meeting of yesterday : http://gnusha.org/lightning-dev/2020-12-07.log, we discussed both #804/#811

Thanks, unfortunately I had a conflict and was only able to log on for the second half! Both of those proposals sound good though.

> What I'm thinking of is landing this PR once minor points are solved and then write down the spec updates to propose changes matching our requirement ?

Sounds good, I'll make these changes tomorrow!

TheBlueMatt commented:
> What I'm thinking of is landing this PR once minor points are solved and then write down the spec updates to propose changes matching our requirement ?

This sounds good to me. Especially given it works in the real world, I see little reason to sit on it forever.

TheBlueMatt reviewed (approved):
LGTM. One note about docs but I'm happy with it.

ariard commented on lightning/src/routing/network_graph.rs:
```diff
@@ -1813,4 +1945,184 @@ mod tests {
+		{
+			let (secp_ctx, net_graph_msg_handler) = create_net_graph_msg_handler();
+			let init_msg = Init { features: InitFeatures::known() };
+			for n in 1..6 {
```
You need to iterate the for-loop until 7, otherwise it doesn't test syncing limitation to FULL_SYNCS_TO_REQUEST ?

Diff:
```
diff --git a/lightning/src/routing/network_graph.rs b/lightning/src/routing/network_graph.rs
index d6c72d08..4b0c2087 100644
--- a/lightning/src/routing/network_graph.rs
+++ b/lightning/src/routing/network_graph.rs
@@ -235,10 +235,10 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
                        return ();
                }
 
-               // Check if we need to perform a full synchronization with this peer
-               if !self.should_request_full_sync(their_node_id) {
-                       return ();
-               }
+               //// Check if we need to perform a full synchronization with this peer
+               //if !self.should_request_full_sync(their_node_id) {
+               //      return ();
+               //}
 
                let first_blocknum = 0;
                let number_of_blocks = 0xffffffff;
```

bmancini55 replied:
Dag-nabbit.  Fixed!  Thanks for catching that.  A minor refactor of the test bit me.

ariard reviewed (commented):
Thanks for the changes, just a small test coverage issue and SGTM :)

ariard commented:
Code Review ACK e0bb63b
